### PR TITLE
Create initial node deployment with the cluster

### LINKED ISSUE #2928: Improve initial node deployment creation
**User Story** 
At the moment when the user is creating a cluster, there are separate requests to create the cluster and to create initial node deployment. Request to create initial node deployment is send from the UI after the user is redirected to the cluster details page and the cluster is healthy. In the meantime, initial node data is stored in the local storage. This causes potential issues, i.e. when the user closes the browser before initial node deployment creation or if the local storage is corrupted/removed.

To improve the situation we can start with changes in the API. We can wait for a cluster to be created in the initial request and create initial node deployment. It shouldn't take too long but it should fix most of the problems.

The target solution for this issue seems to be implementation of a new controller that will handle initial node deployment creation and will store machine deployments/some other CRDs in the seed cluster as a cache/queue and when the clusters will be ready then the controller can add machine deployments into the clusters.

**Acceptance criteria**
- [x] API should handle the creation of initial node deployments.
- [x] There should be a metric and alert for cases where creation will fail after a certain timeout.
- [ ] There should be tests to cover the changes in the API.
- [ ] UI should be adjusted and tested as well.

cc @mrIncompetent @alvaroaleman @kgroschoff @p0lyn0mial @thetechnick   

Feel free to edit this issue if you feel that description could be changed.

### PR DESCRIPTION
**Pull request by maciaszczykm**
**What this PR does / why we need it**: Contains changes that allow creating initial node deployments with the cluster, in the same requests.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: Fixes #2928.

**Special notes for your reviewer**: Has to be merged at the same time as https://github.com/kubermatic/dashboard-v2/pull/1125.

**Documentation**: n/a

**Release note**:

```release-note
Initial node deployments are now created with the cluster.
```

### CODE REVIEW

File: api/pkg/handler/v1/cluster/cluster.go
Comment (p0lyn0mial) [inline api/pkg/handler/v1/cluster/cluster.go]: Checking `Status.Health.AllHealthy()` is not enough. Since we use an impersonated client, RBAC has to be generated before we open a connection the cluster otherwise we might get occasionally 403. 

@mrIncompetent do you think we could extend cluster health to also include RBAC readiness in the cluster ?

Comment (mrIncompetent) [reply api/pkg/handler/v1/cluster/cluster.go]: I think we need such a "checkpoint" on the cluster.

But:
The necessary RBAC permissions are being applied by a controller which has no access to the actual cluster object?

Comment (p0lyn0mial) [reply api/pkg/handler/v1/cluster/cluster.go]: > The necessary RBAC permissions are being applied by a controller which has no access to the actual cluster object?

I think so, the controller (UCCM) runs in `cluster-abc` namespace and adds RBAC to a cluster created by the user.

> I think we need such a "checkpoint" on the cluster.

Not sure if this is the best solution, all I'm saying is that we need a way to know that UCCM has finished his job before we say that the cluster is ready.

Comment (mrIncompetent) [reply api/pkg/handler/v1/cluster/cluster.go]: > Not sure if this is the best solution, all I'm saying is that we need a way to know that UCCM has finished his job before we say that the cluster is ready.

But how do we know its ready?
The 2 potential solutions which come to my mind here:
- Lets the UCCM write a condition in the cluster object when it successfully completed the sync loop
  - This has the downside that we now need to make the UCC "seed cluster aware" which means we need to talk to 2 clusters now. 
- Let the UCCM write some checkpoint resource in the cluster which gets checked by the cluster controller
  - The cluster controller does a "is reachable" check during each sync anyway

Comment (p0lyn0mial) [reply api/pkg/handler/v1/cluster/cluster.go]: since the CM connects to the clusters anyway the second option seems compelling. It could read a "checkpoint" resource and report the status back all the way up to the method that checks if the cluster is healthy.

**Comment by maciaszczykm** [comment]
Added metric to report the number of failures. How do we add them to alert manager?

/hold

**Comment by maciaszczykm** [comment]
PTAL @p0lyn0mial

File: api/pkg/handler/v1/cluster/cluster.go
Comment (mrIncompetent) [reply api/pkg/handler/v1/cluster/cluster.go]: Sounds good, would you mind creating an Issue for that?

File: api/pkg/handler/v1/cluster/cluster.go
Comment (p0lyn0mial) [inline api/pkg/handler/v1/cluster/cluster.go]: we need to add a guard here - to recover from a panic, for example we could use `utilruntime.HandleCrash()` ("k8s.io/apimachinery/pkg/util/runtime) or create our own function.

File: api/pkg/handler/v1/cluster/cluster.go
Comment (p0lyn0mial) [inline api/pkg/handler/v1/cluster/cluster.go]: rm the comment.

File: api/pkg/handler/v1/cluster/cluster.go
Comment (p0lyn0mial) [inline api/pkg/handler/v1/cluster/cluster.go]: do we want to make a `NodeDeployment` optional ? If not provided we create a cluster only.

File: api/pkg/handler/v1/cluster/cluster.go
Comment (p0lyn0mial) [inline api/pkg/handler/v1/cluster/cluster.go]: we could use `wait.Poll`(`"k8s.io/apimachinery/pkg/util/wait"`) method instead, wdyt ?

File: api/pkg/metrics/middleware.go
Comment (p0lyn0mial) [inline api/pkg/metrics/middleware.go]: I would prefer to pass the metric as a dependency to  `CreateEndpoint` function. This file `middleware.go` contains metrics that enhance all HTTP request, whereas we want to add a metric to only one endpoint.

File: api/pkg/metrics/middleware.go
Comment (p0lyn0mial) [inline api/pkg/metrics/middleware.go]: not sure if putting a timeout in the comment is good, it can easily get out of sync with the code.

File: api/pkg/metrics/middleware.go
Comment (p0lyn0mial) [inline api/pkg/metrics/middleware.go]: should we also add a data center name so that we can locate the faulty cluster quicker?

/cc @mrIncompetent

File: api/pkg/metrics/middleware.go
Comment (p0lyn0mial) [inline api/pkg/metrics/middleware.go]: @mrIncompetent  do we want to define an alert for that metric ?

File: api/pkg/resources/machine/machinedeployment.go
Comment (p0lyn0mial) [inline api/pkg/resources/machine/machinedeployment.go]: we usually group imports, add a new line so that they are separated from std imports.

File: api/pkg/resources/machine/machinedeployment.go
Comment (p0lyn0mial) [inline api/pkg/resources/machine/machinedeployment.go]: as below.

File: api/pkg/resources/machine/machinedeployment.go
Comment (p0lyn0mial) [inline api/pkg/resources/machine/machinedeployment.go]: as below.

File: api/pkg/resources/machine/machinedeployment.go
Comment (p0lyn0mial) [inline api/pkg/resources/machine/machinedeployment.go]: it feels weird to return api errors from a function that is not necessarily called from the API server

File: api/pkg/handler/v1/cluster/cluster.go
Comment (p0lyn0mial) [inline api/pkg/handler/v1/cluster/cluster.go]: could we also add an event to the cluster saying that we were unable to add a ND for the cluster? In the future the dashboard will display events for a cluster.

File: api/pkg/handler/v1/cluster/cluster.go
Comment (p0lyn0mial) [inline api/pkg/handler/v1/cluster/cluster.go]: In the worst case scenario, the end user will not receive any status/info back from the API. He might start to think that a ND will not be added to the cluster and could send a new request. I'd add an event saying that a ND is going to be added to the cluster.

**Review by p0lyn0mial — CHANGES_REQUESTED** [review]

File: api/pkg/metrics/middleware.go
Comment (mrIncompetent) [reply api/pkg/metrics/middleware.go]: Adding the seed-dc name would be great!

File: api/pkg/metrics/middleware.go
Comment (mrIncompetent) [reply api/pkg/metrics/middleware.go]: Yes! But we can do that in a follow up in case it would block this PR.
cc @xrstf

File: api/pkg/handler/v1/cluster/cluster.go
Comment (p0lyn0mial) [reply api/pkg/handler/v1/cluster/cluster.go]: sure, np.

File: api/pkg/handler/v1/cluster/cluster.go
Comment (maciaszczykm) [reply api/pkg/handler/v1/cluster/cluster.go]: Makes sense. I will add an issue to track this soon.

Comment (maciaszczykm) [reply api/pkg/handler/v1/cluster/cluster.go]: `this` means adding events related to initial node deployment creation.

File: api/pkg/handler/v1/cluster/cluster.go
Comment (maciaszczykm) [reply api/pkg/handler/v1/cluster/cluster.go]: Yes, that's why there is a `TODO` above. Would you like to have it as a part of this PR?

File: api/pkg/resources/machine/machinedeployment.go
Comment (maciaszczykm) [reply api/pkg/resources/machine/machinedeployment.go]: True, I will change it.

File: api/pkg/handler/v1/cluster/cluster.go
Comment (p0lyn0mial) [reply api/pkg/handler/v1/cluster/cluster.go]: ah :) Your call :) I'm fine with a follow-up.

File: api/pkg/handler/v1/cluster/cluster.go
Comment (p0lyn0mial) [reply api/pkg/handler/v1/cluster/cluster.go]: added an issue https://app.zenhub.com/workspaces/all-kubermatic-58863654794153994b62af76/issues/kubermatic/kubermatic/3019

File: api/pkg/handler/v1/cluster/cluster.go
Comment (maciaszczykm) [reply api/pkg/handler/v1/cluster/cluster.go]: https://github.com/kubermatic/kubermatic/issues/3027

File: api/pkg/handler/v1/cluster/cluster.go
Comment (maciaszczykm) [reply api/pkg/handler/v1/cluster/cluster.go]: https://github.com/kubermatic/kubermatic/issues/3027

File: api/pkg/handler/v1/cluster/cluster.go
Comment (maciaszczykm) [reply api/pkg/handler/v1/cluster/cluster.go]: Yep, could be useful for the future.

**Comment by maciaszczykm** [comment]
PTAL @p0lyn0mial

**Comment by maciaszczykm** [comment]
If it will be approved I will try to merge in the evening and then UI part as soon as e2e will pass.

File: api/pkg/handler/routes_v1.go
Comment (p0lyn0mial) [inline api/pkg/handler/routes_v1.go]: grouping :(

File: api/pkg/handler/v1/cluster/cluster.go
Comment (p0lyn0mial) [inline api/pkg/handler/v1/cluster/cluster.go]: we could avoid repeating calling `initNodeDeploymentFailures` in `createInitialNodeDeployment` function by doing:
```
go func() {
 defer utilruntime.HandleCrash()
 err := createInitialNodeDeployment(...)
 if err != . il {
   initNodeDeploymentFailures.With(prometheus.Labels{"cluster": cluster.Name, "seed_dc": seedDC}).Add(1)
 }
}()
```

what do you think ?

File: api/cmd/kubermatic-api/metrics.go
Comment (p0lyn0mial) [inline api/cmd/kubermatic-api/metrics.go:26]: could you add an item for creating an alert for the metric ?

File: api/pkg/handler/routes_v1.go
Comment (maciaszczykm) [reply api/pkg/handler/routes_v1.go]: Sorry... Is there any check for GoLand for it?

Comment (p0lyn0mial) [reply api/pkg/handler/routes_v1.go]: I'm afraid there is not, let me know once you find one :)

File: api/cmd/kubermatic-api/metrics.go
Comment (maciaszczykm) [reply api/cmd/kubermatic-api/metrics.go:26]: https://github.com/kubermatic/kubermatic/issues/3041

File: api/pkg/handler/v1/cluster/cluster.go
Comment (maciaszczykm) [reply api/pkg/handler/v1/cluster/cluster.go]: Awesome, it improves a lot :)

**Comment by maciaszczykm** [comment]
PTAL @p0lyn0mial

**Comment by p0lyn0mial** [comment]
this lgtm waiting for @mrIncompetent before approval

thanks for the pull :)

File: api/cmd/kubermatic-api/metrics.go
Comment (mrIncompetent) [inline api/cmd/kubermatic-api/metrics.go]: Add `kubermatic_api` to all metric names

File: api/cmd/kubermatic-api/metrics.go
Comment (mrIncompetent) [inline api/cmd/kubermatic-api/metrics.go:39]: For gauge & counter metrics: Initialize them with 0 within the register func:
For example: `metrics.HTTPRequestsTotal.Add(0)`

File: api/cmd/kubermatic-api/metrics.go
Comment (mrIncompetent) [inline api/cmd/kubermatic-api/metrics.go:1]: Why not just declare them in the package where they actually get used?

File: api/pkg/handler/test/hack/hack.go
Comment (mrIncompetent) [inline api/pkg/handler/test/hack/hack.go:84]: Why only partially initializes this ~metric~ struct?

File: api/cmd/kubermatic-api/metrics.go
Comment (maciaszczykm) [reply api/cmd/kubermatic-api/metrics.go:1]: We wanted to group API's metrics in one file. Isn't that good?

cc @p0lyn0mial

File: api/pkg/handler/v1/cluster/cluster.go
Comment (mrIncompetent) [inline api/pkg/handler/v1/cluster/cluster.go]: Can we increase the timeout to 30min ?
If you have new worker nodes which need to pull all images, there's a good chance that it will take >5min

Comment (maciaszczykm) [reply api/pkg/handler/v1/cluster/cluster.go]: Of course, I will do it.

File: api/pkg/handler/test/hack/hack.go
Comment (maciaszczykm) [reply api/pkg/handler/test/hack/hack.go:84]: So basically having this init inside a function where we also register the metrics and set counters to 0?

File: api/cmd/kubermatic-api/metrics.go
Comment (mrIncompetent) [reply api/cmd/kubermatic-api/metrics.go:1]: I prefer to just add the metrics in the packages where they are actually needed & register them in an init func.
Like: https://github.com/kubermatic/kubermatic/blob/master/api/pkg/controller/cluster/metrics.go

Then you don't need to do any pluming - You now have a hard dependency and there's no easy way to unit-test those, but its easier from my point of view.
But that's just my opinion - this don't must be changed

File: api/pkg/handler/v1/cluster/cluster.go
Comment (p0lyn0mial) [reply api/pkg/handler/v1/cluster/cluster.go]: the timeout (5 min) is for waiting for a cluster to become ready, we are not waiting for worker nodes.

File: api/pkg/handler/test/hack/hack.go
Comment (mrIncompetent) [reply api/pkg/handler/test/hack/hack.go:84]: Ah wait, i'm an idiot. That's test code. Sorry :man_facepalming:

File: api/pkg/handler/v1/cluster/cluster.go
Comment (mrIncompetent) [reply api/pkg/handler/v1/cluster/cluster.go]: > the timeout (5 min) is for waiting for a cluster to become ready, we are not waiting for worker nodes.

Misunderstanding:
If the worker nodes of the seed cluster are "fresh" and need to pull the images for the control plane pods, this step can take definitely more than 5min.

File: api/cmd/kubermatic-api/metrics.go
Comment (p0lyn0mial) [reply api/cmd/kubermatic-api/metrics.go]: make sure to update alerts when the name changes.

File: api/pkg/handler/v1/cluster/cluster.go
Comment (maciaszczykm) [reply api/pkg/handler/v1/cluster/cluster.go]: > the timeout (5 min) is for waiting for a cluster to become ready, we are not waiting for worker nodes.

True, perhaps 10 minutes will be a compromise here.

Comment (p0lyn0mial) [reply api/pkg/handler/v1/cluster/cluster.go]: oh, I see, they need to download Kubernetes images. Thanks.

File: api/cmd/kubermatic-api/metrics.go
Comment (maciaszczykm) [reply api/cmd/kubermatic-api/metrics.go:39]: Hmm, what to do with labels? Leave them empty? Seems like there is no Add() method directly on the metric object.

File: api/pkg/handler/v1/cluster/cluster.go
Comment (maciaszczykm) [reply api/pkg/handler/v1/cluster/cluster.go]: Okay, then 30m it is 🙂

File: api/cmd/kubermatic-api/metrics.go
Comment (p0lyn0mial) [reply api/cmd/kubermatic-api/metrics.go:1]: no strong opinion here, I figured it would be nice to declare them in the main pkg as it improves readability - you immediately know what type of metrics are collected by the API server.

File: api/cmd/kubermatic-api/metrics.go
Comment (mrIncompetent) [reply api/cmd/kubermatic-api/metrics.go:39]: Ah sh*t.
Haven't noticed that. So my comment is useless, sorry!

File: api/cmd/kubermatic-api/metrics.go
Comment (maciaszczykm) [reply api/cmd/kubermatic-api/metrics.go]: Yep, I did.

File: api/pkg/handler/v1/cluster/cluster.go
Comment (maciaszczykm) [reply api/pkg/handler/v1/cluster/cluster.go]: Done.

File: api/cmd/kubermatic-api/metrics.go
Comment (maciaszczykm) [reply api/cmd/kubermatic-api/metrics.go:39]: Okay, thanks, no problem 🙂

**Comment by p0lyn0mial** [comment]
thanks for reviewing 

/lgtm
/approve

File: api/cmd/kubermatic-api/metrics.go
Comment (maciaszczykm) [reply api/cmd/kubermatic-api/metrics.go:1]: 3 x no strong opinion 😄 

What about keeping it w/o changes and updating if it won't work for us in the long run?

**Comment by maciaszczykm** [comment]
Thank you both for review and tips :)

**Comment by p0lyn0mial** [comment]
@kron4eg could you please tag the pull ?

/hold cancel

File: config/monitoring/prometheus/rules/kubermatic-master-kubermatic.yaml
Comment (kron4eg) [inline config/monitoring/prometheus/rules/kubermatic-master-kubermatic.yaml]: * this file is generated (there is a notice in the head: "# This file has been generated, do not edit.")
* renaming `http_requests_total` to `kubermatic_api_http_requests_total` is bad idea in general

File: config/monitoring/prometheus/rules/src/kubermatic-master/kubermatic.yaml
Comment (kron4eg) [inline config/monitoring/prometheus/rules/src/kubermatic-master/kubermatic.yaml]: same applies here

File: config/monitoring/prometheus/rules/kubermatic-master-kubermatic.yaml
Comment (p0lyn0mial) [reply config/monitoring/prometheus/rules/kubermatic-master-kubermatic.yaml]: could you say why renaming is a bad idea ?

Comment (kron4eg) [reply config/monitoring/prometheus/rules/kubermatic-master-kubermatic.yaml]: * metrics names is public API and need to be threaded the same way
* this breaks ability to aggregate across all `http_requests_total` from all the services
* it adds nothing (we were able to filter only to kubermatic `http_requests_total{role="kubermatic-api"}` previously just fine)

Comment (maciaszczykm) [reply config/monitoring/prometheus/rules/kubermatic-master-kubermatic.yaml]: @mrIncompetent are you okay with reverting this?

Comment (mrIncompetent) [reply config/monitoring/prometheus/rules/kubermatic-master-kubermatic.yaml]: I'm basically just following https://prometheus.io/docs/practices/naming/#metric-names

> A metric name should have a (single-word) application prefix relevant to the domain the metric belongs to. The prefix is sometimes referred to as namespace by client libraries. For metrics specific to an application, the prefix is usually the application name itself. Sometimes, however, metrics are more generic, like standardized metrics exported by client libraries

Comment (mrIncompetent) [reply config/monitoring/prometheus/rules/kubermatic-master-kubermatic.yaml]: But apparently i misinterpreted this one: https://groups.google.com/forum/#!topic/prometheus-developers/XjlOnDCK9qc

Comment (kron4eg) [reply config/monitoring/prometheus/rules/kubermatic-master-kubermatic.yaml]: yeah,
>Sometimes, however, metrics are more generic, like standardized metrics exported by client libraries

`http_requests_total` is one of those standard metrics that a useful when unified across different services.

**Comment by maciaszczykm** [comment]
Removed the prefix from the metrics names. @kron4eg PTAL.

File: api/cmd/kubermatic-api/metrics.go
Comment (kron4eg) [reply api/cmd/kubermatic-api/metrics.go]: This metric is custom, and probably should get new prefixed name per @mrIncompetent comment.

**Comment by maciaszczykm** [comment]
@kron4eg PTAL

**Review by kron4eg — APPROVED** [review]
/lgtm

**Pull request merged**
